### PR TITLE
gh-103092: Isolate _collections

### PR DESCRIPTION
**Pull request by erlend-aasland**
* Issue: gh-103092

### CODE REVIEW

**Review by kumaraditya303 — APPROVED** [review]
LGTM, I finished the remaining bits.

**Comment by kumaraditya303** [comment]
Now it crashes in the dict watcher deallocation, cc @carljm

**Comment by erlend-aasland** [comment]
> Now it crashes in the dict watcher deallocation, cc @carljm

See 45a26f3bed971d9d8a08bdf293e96f97b50dcabe, which I believe is the correct prescription. Notice that `PyDict_Type.tp_init` is called explicitly from `defdict_init`. IIRC, that is not needed, but I may be wrong (I'm unable to check that right now).

**Comment by erlend-aasland** [comment]
I've fixed the type names (fixes pickle, pickletools, and dataclasses tests). I need to fix up `defdict_or` to fix `test_defaultdict`; current code is wrong and assumes it can derive the state from the left argument, but we need to check both args. WIP.

File: Modules/_collectionsmodule.c [hunk 1]
```diff
@@ -2112,12 +2112,24 @@ static PyObject*
 defdict_or(PyObject* left, PyObject* right)
 {
     PyObject *self, *other;
-    collections_state *state = find_module_state_by_def(Py_TYPE(left));
+
+    // Find module state
```
Comment (kumaraditya303) [inline Modules/_collectionsmodule.c:2116]: For my own reference this happens because `slot_nb_or` from typeobject.

**Review by kumaraditya303 — APPROVED** [review]

**Comment by erlend-aasland** [comment]
BTW, there are ref. leaks, so we need to hold off a little bit more. Try for example checking ref. leaks on `test_defaultdict`.

**Comment by erlend-aasland** [comment]
Good find! No ref.leaks locally over here.

**Review by kumaraditya303 — APPROVED** [review]
Final approval!

**Comment by erlend-aasland** [comment]
Good job!

**Pull request merged**